modules/ding_availability/js/ding_availability_labels.js

```diff
@@ -9,205 +9,193 @@
+      group = $('.js-reservable', groups_wrapper);
+      if (group.length === 0) {
+        group = $('<p class="js-reservable">' + Drupal.t('Reservable') + ': </p>');
+        groups_wrapper.append(group);
+      }
+    }
+    else {
+      group = $('.js-unavailable', groups_wrapper);
+
+      if (group.length === 0) {
+        group = $('<p class="js-unavailable">' + Drupal.t('Not available') + ': </p>');
+        groups_wrapper.append(group);
+      }
+    }
+
+    // Move the element into that type.
+    group.append(element);
+
+    // Remove empty groups.
+    $('.js-available, .js-reservable, .js-unavailable', groups_wrapper)
+      .not(':has(.js-search-result--availability-link)')
+      .remove();
+  }
+
+  /**
+   * Removes js-pending groups (labels) if they are empty or changes the
+   * label to "Can be obtained". This should be called as the last function
+   * in updating availability information and see as a clean-up function.
+   */
+  function ding_availability_labels_remove_pending() {
+    // Loop over all pending availability groups.
+    $('.js-pending').each(function() {
+      var elm = $(this);
+      var children = elm.children();
+      if (children.length) {
+        // Change the label from pending.
+        var label = elm.contents().first()[0];
+        label.nodeValue = Drupal.t('Can be obtained') + ': ';
+      }
+      else {
+        // The current pending group is empty, so simply remove it.
+        elm.remove();
+      }
+    });
+  }
+
+  /**
+   * Attach the availability behaviors to the page.
+   *
+   * The will be re-attached at every page content update.
+   */
+  Drupal.behaviors.ding_availability_labels = {
     attach: function(context, settings) {
       var ids = [];
       var html_ids = [];
 
-      // Extract entity ids from Drupal settings array.
+      // Loop through the materials given in the settings and collect
+      // id.
       if (settings.hasOwnProperty('ding_availability')) {
-        $.each(settings.ding_availability, function(id, entity_ids) {
-          $.each(entity_ids, function(index, entity_id) {
-            if (Drupal.DADB[entity_id] === undefined) {
-              Drupal.DADB[entity_id] = null;
-              ids.push(entity_id);
-              html_ids.push(id);
-            }
-          });
+        $.each(settings.ding_availability, function (id, data) {
+          if (!Drupal.DADB.hasOwnProperty(data.html_id)) {
+            Drupal.DADB[data.html_id] = null;
+            ids[data.html_id] = data;
+            html_ids.push(data.html_id);
+          }
         });
       }
 
       // Fetch availability.
-      if (ids.length > 0) {
-        var mode = settings.ding_availability_mode ? settings.ding_availability_mode : 'items';
-        var path = settings.basePath + 'ding_availability/' + mode + '/' + ids.join(',');
+      if (!$.isEmptyObject(ids)) {
+        var path = settings.basePath + 'ding_availability/items';
         $.ajax({
+          type: "POST",
```

Comment:
This will prevent default caching of availability requests - although these bypass Varnish anyhow at the moment.

Reply:
I do not think it is the responsibility of the client to decide if the server can cache the response or not. `POST` requests are intended for requests which change state. Checking for availability does not. If such a request should not be cached then it should be handled by sending the proper response headers instead.

Reply:
I do not think it is the responsibility of the client to decide if the server can cache the response or not. POST requests are intended for requests which change state. Checking for availability does not. If such a request should not be cached then it should be handled by sending the proper response headers instead.